### ISSUE TITLE
Generating the same sentence embedding for `all-MiniLM-L6-v2` using `candle`

### ISSUE THREAD

**Issue by radu-matei** [huggingface/candle#380]
https://github.com/huggingface/candle/blob/25ec2d9f6bf36ff51c04f54f6c243828f6f4a8da/candle-examples/examples/bert/main.rs#L164-L167

The BERT sentence embedding example is using a pooling strategy that generates a different sentence embedding compared to using either the HuggingFace API, or alternative ways of running the model locally.

I would be interested in getting the same result, and I suspect it's in the pooling strategy that should be used.

Any pointers would be helpful.

Thanks!

**Comment by LaurentMazare** [comment]
Thanks for reporting this. Would you mind providing some python and rust snippet of the code that you ran so that it's easier to reproduce the issue on our side?

**Comment by radu-matei** [comment]
The code above generates the embeddings using `candle`.

The simplest way to compare it with the expected output is by using the HF API:

```
$ curl https://api-inference.huggingface.co/pipeline/feature-extraction/sentence-transformers/all-MiniLM-L6-v2 -X POST -H "Authorization: Bearer hf_XXX" -d "Sample sentence"
```

**Comment by krolinventions** [comment]
Did you normalize the embedding? When I did that it candle matched the output from rust-bert exactly.

**Comment by kaxap** [comment]
couldn't get results to match even after the normalization. Could you please provide an example of the normalization using Candle?

**Comment by krolinventions** [comment]
Sure, this is the code I'm using:

```
        let token_ids = Tensor::stack(&token_ids, 0)?;
        let token_type_ids = token_ids.zeros_like()?;
        let embeddings = self.model.forward(&token_ids, &token_type_ids)?;

        // Apply some avg-pooling by taking the mean embedding value for all tokens (including padding)
        let (_n_sentence, n_tokens, _hidden_size) = embeddings.dims3()?;
        let embeddings = (embeddings.sum(1)? / (n_tokens as f64))?;

        let mut results: Vec<Vec<f32>> = Vec::new();
        for j in 0..n_sentences {
            let e_j = embeddings.get(j)?;
            let mut emb: Vec<f32> = e_j.to_vec1()?;
            normalize(&mut emb);
            results.push(emb);
        }
```

With `normalize` defined as:

```
pub fn normalize(v: &mut Vec<f32>) {
    let length: f32 = v.iter().map(|x| x * x).sum::<f32>().sqrt();
    v.iter_mut().for_each(|x| *x /= length);
}
```

See https://github.com/dawn-search/dawnsearch/blob/main/src/embedding/embedding_service.rs

**Comment by radu-matei** [comment]
After normalizing the embeddings, I am getting consistent results across the implementations.

Thanks!

**Comment by krolinventions** [comment]
I think it would be good to update the example to use normalized embeddings, as I think that's what most people will be using. I can give that a try later.

**Comment by LaurentMazare** [comment]
I've just merged #390 which adds some optional embedding normalization to the bert example, please let us know if you still see some discrepencies.
